substrate/client/rpc-spec-v2/src/common/connections.rs

```diff
@@ -0,0 +1,258 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use jsonrpsee::ConnectionId;
+use parking_lot::Mutex;
+use std::{
+	collections::{HashMap, HashSet},
+	sync::Arc,
+};
+
+/// Connection state which keeps track whether a connection exist and
+/// the number of concurrent operations.
+#[derive(Default, Clone)]
+pub struct RpcConnections {
+	/// The number of identifiers that can be registered for each connection.
+	///
+	/// # Example
+	///
+	/// This is used to limit how many `chainHead_follow` subscriptions are active at one time.
+	capacity: usize,
+	/// Map the connecton ID to a set of identifiers.
+	data: Arc<Mutex<HashMap<ConnectionId, ConnectionData>>>,
+}
+
+#[derive(Default)]
+struct ConnectionData {
+	/// The total number of identifiers for the given connection.
+	///
+	/// An identifier for a connection might be:
+	/// - the subscription ID for chainHead_follow
+	/// - the operation ID for the transactionBroadcast API
+	/// - or simply how many times the transaction API has been called.
+	///
+	/// # Note
+	///
+	/// Because a pending subscription sink does not expose the future subscription ID,
+	/// we cannot register a subscription ID before the pending subscription is accepted.
+	/// This variable ensures that we have enough capacity to register an identifier, after
+	/// the subscription is accepted. Otherwise, a jsonrpc error object should be returned.
+	num_identifiers: usize,
+	/// Active registered identifiers for the given connection.
+	///
+	/// # Note
+	///
+	/// For chainHead, this represents the subscription ID.
+	/// For transactionBroadcast, this represents the operation ID.
+	/// For transaction, this is empty and the number of active calls is tracked by
+	/// [`Self::num_identifiers`].
+	identifiers: HashSet<String>,
```

Comment:
Am I right to assume that we are not worries about subscription IDs and operation IDs colliding at all in our implementation here?

Reply:
Yep, IIRC, we'll either use this for chainHead with subscriptions ID only; or for the tx with operation ID only